### ISSUE TITLE
[HOLD for payment 2023-05-23] [$1000] image changes automatic when image preview is opened and sender sends any new image

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#17743]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. click on + button and select new chat
2. create a new chat
3. from userA, send an image 
4. for userB, click on that image to see full preview
5. from userA, send second image
6. check  image preview on userB side

## Expected Result: 
Image should not changes automatically
## Actual Result:
Image changes automatically

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.2
**Reproducible in staging?:** y
**Reproducible in production?:** y
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/233485002-555426ae-c4bd-4977-a0f2-9647c4db3476.mov

https://user-images.githubusercontent.com/43996225/233485020-4553d93e-8234-4499-a654-246aa0173e83.mp4

**Issue reported by:** @gadhiyamanan 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1681972760701989

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01f32126fc3213fe91</li>
        <li>Upwork Job ID: 1649466254320713728</li>
        <li>Last Price Increase: 2023-05-08</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @JmillsExpensify (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by JmillsExpensify** [comment]
I'm not able to reproduce this one. I followed the reproduction steps using both file upload as well as drag and drop. The preview didn't switch on me. Mind confirming the bug can reproduced using these exact reproduction steps?

**Comment by gadhiyamanan** [comment]
@JmillsExpensify Please try to create a new chat, I am able to reproduce it

https://user-images.githubusercontent.com/54790231/233543073-fde53f41-964c-40a3-aa2e-592b781c6567.mp4

**Comment by JmillsExpensify** [comment]
Ah ok! So creating a **new** chat does it for me as well. Thanks!

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~01f32126fc3213fe91

**Comment by MelvinBot** [comment]
Current assignee @JmillsExpensify is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @sobitneupane (`External`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @cead22 (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by akinwale** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The current image being displayed in the attachment carousel automatically changes when a new image attachment is added to the report.

### What is the root cause of that problem?
There are multiple things causing with the implemented logic in the `AttachmentCarousel` component.

In the `makeStateWithReports` method, the `getSortedReportActions` method is called with the messages sorted in descending order, however, attachments are being added to the resulting array with logic that assumes that the messages are sorted in ascending order.

https://github.com/Expensify/App/blob/4947508bedc2e901082d805c84498f5f394ef655/src/components/AttachmentCarousel/index.js#L124

https://github.com/Expensify/App/blob/4947508bedc2e901082d805c84498f5f394ef655/src/components/AttachmentCarousel/index.js#L148

When this happens on subsequent updates, the order of the attachments array changes, because the second image in the report now becomes the first in order (descending order), thereby causing the `FlatList` to re-render and the new image is displayed.

### What changes do you think we should make in order to solve the problem?
Change the `getSortedReportActions` method call to sort the items in ascending order, then swap the checks for `isForwardDisabled` and `isBackDisabled`, and swap the parameters for the `this.cycleThroughAttachments` calls.

### What alternative solutions did you explore? (Optional)
None.

**Comment by kidroca** [comment]
@allroundexperts 
`onViewableItemsChanged` is called because the visible image changes - we can observe the image changes even before the actual callback is triggered 

The image changes because we insert an image in the list 
The list knows it's at a given index, like 5. If when new attachment is added we insert it to the front (at index 0) then the images would shift up and so the image at index 5 would no longer be the same, because the image at index five now has a different `key` the callback `onViewableItemsChanged` is triggered

@akinwale If you reverse the list, you would have the same problem, just on the other side of the list

**Comment by akinwale** [comment]
> @akinwale If you reverse the list, you would have the same problem, just on the other side of the list

I haven't seen the problem exhibit itself, and I don't think it happens if the list order always remains the same. We're always starting from the first item in the report, and so the page index always stays the same for the current item (if there's any).

**Comment by kidroca** [comment]
### Proposal
#### Please re-state the problem that we are trying to solve in this issue.
Observed attachment image might change if someone sends a new attachment while we have the attachment view/carousel open

#### What is the root cause of that problem?
The problem is we try to recalculate attachments and fit the new attachment into the list - why do we do that?
Imagine the reverse scenario - someone deletes an attachment from the chat while we're looking at it - what do we expect to happen 
- do we expect to be on a different image?
- or do we expect to be able to see the deleted attachment until we close the modal?

#### What changes do you think we should make in order to solve the problem?
A solution like stop trying to recalculate attachments after the carousel is open is simple and efficient 
If attachment comments are added/deleted while the attachment view is open it should simply do nothing, because
- it's bad UX to be kicked out of (lose) an attachment while you're still looking at it
- new messages (adding attachments) are **unread**, having a new attachment added to the carousel, but not being able to distinguish it as something new is bad UX, when we receive new messages we can visually distinguish them as new - there a new message indicator, in the context of the carousel there's no such thing - it would be better to exit the carousel and see what's new, instead of having it appended there automatically

The carousel should include all the attachments available at the time when we open the carousel 
It should not include attachment from unread messages, because the images can't be distinguished as new
New attachments would be added once we exit, read the new message and click to see the attachment (open the carousel again)

#### What alternative solutions did you explore? (Optional)

In my previous comment I've explained how content in the list might shift depending on where we insert data 
If content is to change dynamically so that an image at index 5 is moved to index 6, and we're currently looking at that image, we should tell the list to scroll to index 6, when we're done with the recalculation of attachments

Simply said, any time we use the `makeStateWithReports` we should also tell the list to scroll to the calculated page index

_(This should also cover images getting deleted and causing indexes to change, though if we were looking at a image that got deleted, we'll be moved to the neighboring image)_

**Comment by sobitneupane** [comment]
Here are two cases:

- When user deletes a photo
- When user sends a new photo

I believe when user deletes an attachment, it should be removed from carousel as well (same is the case in WhatsApp as well). When user sends a new attachment, should it be accessible without closing the carousel and re-opening it? 

@kidroca You are proposing not to remove the deleted image from the carousel until it is closed, right? It is not the case in WhatsApp. I think it makes sense to remove the image from the carousel once deleted.

@cead22 @JmillsExpensify What's your opinion on the expected output?

**Comment by akinwale** [comment]
> Here are two cases:
> 
> * When user deletes a photo
> * When user sends a new photo
> 
> I believe when user deletes an attachment, it should be removed from carousel as well (same is the case in WhatsApp as well). When user sends a new attachment, should it be accessible without closing the carousel and re-opening it?
> 

For deletion, this will be a bit tricky if the image is the currently displayed image, because it will cause the `FlatList` to refresh, thereby resulting in confusion for a user who may be viewing the carousel at the time.

If the intent is to dynamically handle deletion, a good approach would be to only update carousel if:
1. The deleted image is not the current image, and
2. The deleted image has a higher index in the attachments list than the current image, preventing a reorder of page indexes which may cause the `FlatList` to flicker or refresh (this needs to be tested some more).

My proposal fixes the issue with dynamically added images changing the current displayed image fine, but not deletion. I can run some tests and update it to cover the deletion cases.

**Comment by akinwale** [comment]
@sobitneupane Another thing to note, at the moment, the current behaviour when an image attachment is deleted is that no changes happen in the carousel while it's already open.

**Comment by JmillsExpensify** [comment]
>Another thing to note, at the moment, the current behaviour when an image attachment is deleted is that no changes happen in the carousel while it's already open.

I'm not particularly passionate, though I agree that if we need to refresh the carousel for the case of deleted, I think that'd cause more confusion by the user – or perhaps more accurately, is jarring.

**Comment by JmillsExpensify** [comment]
@sobitneupane Do you need anything else to help review this proposal?

**Comment by sobitneupane** [comment]
Oops Missed it. 

@JmillsExpensify I am still not clear about expected output. https://github.com/Expensify/App/issues/17743#issuecomment-1519474117

Is following the expected output?

- When user deletes a photo:
The deleted photo will not be removed from already opened carousel.

- When user sends a new photo
The newly sent photo will not be shown in already opened carousel.

**Comment by cead22** [comment]
That's my understanding 👍

**Comment by MelvinBot** [comment]
📣 It's been a week! Do we have any satisfactory proposals yet? Do we need to adjust the bounty for this issue? 💸

**Comment by cead22** [comment]
> A solution, like stop trying to recalculate attachments after the carousel is open, is simple and efficient

@sobitneupane while I think @kidroca 's proposal could use a little more details about how it needs to be implemented, I like this approach, what do you think?

**Comment by sobitneupane** [comment]
@cead22 Yep. I agree. [The proposal](https://github.com/Expensify/App/issues/17743#issuecomment-1518644347) has explained well about what should be the expected output. More details on the solution will be helpful.

**Comment by kidroca** [comment]
> > A solution, like stop trying to recalculate attachments after the carousel is open, is simple and efficient
> 
> @sobitneupane while I think @kidroca 's proposal could use a little more details about how it needs to be implemented, I like this approach, what do you think?

Actually it's quite simple to achieve what I've proposed

In the Carousel we have code that runs a remapping every times `comments` changed
https://github.com/Expensify/App/blob/d60f21ca5f7000c6e1a51661269330d82eccafaa/src/components/AttachmentCarousel/index.js#L175-L178

This remapping is used for initial mapping of attachments and then called in `componentDidUpdate` 
My proposal is to stop using it in `componentDidUpdate` 
https://github.com/Expensify/App/blob/d60f21ca5f7000c6e1a51661269330d82eccafaa/src/components/AttachmentCarousel/index.js#L79-L86

This way actions changing the report comments would not be applied immediately on the carousel, but on the next time we open it

**Comment by cead22** [comment]
That makes sense. Why was `this.makeStateWithReports();` added to `componentDidUpdate` in the first place / what function does it serve now?

**Comment by kidroca** [comment]
> That makes sense. Why was `this.makeStateWithReports();` added to `componentDidUpdate` in the first place / what function does it serve now?

It is the method that filters out comments with attachments. It serves to prepare a list of attachments rendered in the carousel

It was added to `componentDidUpdate` with the original code that created the component

I'm not sure why it was decided to be this way

**Comment by cead22** [comment]
Thanks! @kidroca 's proposal looks simpler to me, @sobitneupane what do you think?

**Comment by sobitneupane** [comment]
Thanks @kidroca. It will solve the issue on addition of new image. But when user deletes an image, the carousel closes abruptly which is not expected.

**Comment by kidroca** [comment]
> Thanks @kidroca. It will solve the issue on addition of new image. But when user deletes an image, the carousel closes abruptly which is not expected.

I didn't see this in the issue definition.

I don't think this bug is related to the code I'm suggesting to change.

When does it happen 
- now 
- or with my proposal?

**Comment by sobitneupane** [comment]
It is happening now. And I believe we want it to be addressed here. In fact you started the discussion about it in [your proposal](https://github.com/Expensify/App/issues/17743#issuecomment-1518644347). 

> A solution, like stop trying to recalculate attachments after the carousel is open, is simple and efficient
If attachment comments are added/deleted while the attachment view is open it should simply do nothing, because
> - it's bad UX to be kicked out of (lose) an attachment while you're still looking at it

Expected output discussed [here](https://github.com/Expensify/App/issues/17743#issuecomment-1527516066).

**Comment by kidroca** [comment]
I've dug around a bit and found out the cause for the "Carousel gets closed bug" - it's because the modal that renders the carousel is a child of the thumbnail image in the chat comment. So when that comment gets deleted all it's child rendered elements are removed as well 

Here's the HTML renderer that translates images to React components. It renders `<AttachmentModal>` which renders the carousel
https://github.com/Expensify/App/blob/7dd86efc25d4c9c0f72563bf5e8eece353ab3e16/src/components/HTMLEngineProvider/HTMLRenderers/ImageRenderer.js#L61

This is a separate issue and requires a different approach

**Comment by cead22** [comment]
I agree, and was able to reproduce that issue of the modal closing for the receiving user when the message with the attachment is deleted by the sending user.

@sobitneupane are you good with @kidroca 's proposal?

**Comment by sobitneupane** [comment]
Yep @cead22. We can go with @kidroca's proposal to handle this specific issue. We will have to create another issue to handle attachment deletion case.

**Comment by kidroca** [comment]
I reported it and it was created here
- https://github.com/Expensify/App/issues/18451

**Comment by cead22** [comment]
Great, thanks! @JmillsExpensify can you do the needful please?

**Comment by cead22** [comment]
Not overdue. Assigned @kidroca

**Comment by JmillsExpensify** [comment]
@kidroca Upwork job is here when you're ready: https://www.upwork.com/jobs/~01f32126fc3213fe91.

**Comment by kidroca** [comment]
Pull request ready for review
- https://github.com/Expensify/App/pull/18651

**Comment by cead22** [comment]
The PR for this is on staging

**Comment by JmillsExpensify** [comment]
Re-applying the `bug` label, as I'm OOO the rest of this week.

**Comment by miljakljajic** [comment]
Also OOO, reapplying.

**Comment by laurenreidexpensify** [comment]
@sobitneupane @kidroca @gadhiyamanan the original job expired on Upwork, I've resent you all offers on a new one - please apply here https://www.upwork.com/jobs/~01b269c8c394133c7a thanks

**Comment by laurenreidexpensify** [comment]
@sobitneupane have you had a chance to review regression steps above?

**Comment by laurenreidexpensify** [comment]
@sobitneupane @gadhiyamanan paid ✔️
waiting for @kidroca to accept job in upwork for final payment

**Comment by sobitneupane** [comment]
> **BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:
> 
> * [ ]  [@sobitneupane] The PR that introduced the bug has been identified. Link to the PR:

https://github.com/Expensify/App/pull/9279
> * [ ]  [@sobitneupane] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:

https://github.com/Expensify/App/pull/9279#issuecomment-1560708427

> * [ ]  [@sobitneupane] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:

https://expensify.slack.com/archives/C049HHMV9SM/p1684918537590419

**Comment by sobitneupane** [comment]
**Regression Test Proposal**

1. Set up a chat between two users (User A and User B).
2. Have User A open the carousel and navigate to the very first image.
3. Have User B send a new attachment.
4. Verify that the currently viewed image in the carousel by User A remains the same.

Do we agree 👍 or 👎

**Comment by kidroca** [comment]
@laurenreidexpensify am I eligable for the bonus mentioned here: https://github.com/Expensify/App/issues/17743#issuecomment-1550460943
> As a reminder, here are the bonuses/penalties that should be applied for any External issue:

> Merged PR within 3 business days of assignment - 50% bonus
> Merged PR more than 9 business days after assignment - 50% penalty

- I was assigned for the ticket on May 9th - https://github.com/Expensify/App/issues/17743#issuecomment-1539186168
- Create a PR which was merged on May 11th - https://github.com/Expensify/App/pull/18651#issuecomment-1544278407

The PR was merged within 72hrs of my assignment

**Comment by laurenreidexpensify** [comment]
sorry missed the bonus @kidroca @sobitneupane - fixed!

**Comment by sobitneupane** [comment]
@laurenreidexpensify I am not familiar with TestRail. Is there any document/ walkthrough video/ slack thread to familiarize with TestRail?

**Comment by laurenreidexpensify** [comment]
ah apols Sobit - Reading through https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md I can see that you don't have access to TestRail, so ignore me! 

Testing steps look good, have created an issue for it to get added to TestRail. 

All steps completed here, closing. Thanks everyone!
